### PR TITLE
Feature/KYV-840 mit charge subscription renewal

### PR DESCRIPTION
**Pull request by wpamb**

### CODE REVIEW

**Review by laukkanenj — APPROVED** [review]

File: paymentapi/src/main/java/fi/hel/verkkokauppa/payment/api/PaymentAdminController.java [hunk 1]
```diff
@@ -100,6 +102,45 @@ public ResponseEntity<List<Payment>> getPayments(@RequestParam(value = "orderId"
...
+                log.warn("paid payment exists, not creating new payment for orderId: " + message.getOrderId());
+            } else {
+                if (Boolean.TRUE.equals(message.getIsSubscriptionRenewalOrder()) && message.getCardToken() != null && message.getCardExpYear() != null && message.getCardExpMonth() != null && message.getCardLastFourDigits() != null) {
+                    Payment payment = onlinePaymentService.createSubscriptionRenewalPayment(message);
+                    PaymentCardInfoDto card = new PaymentCardInfoDto(message.getCardToken(), message.getCardExpYear(), message.getCardExpMonth(), message.getCardLastFourDigits());
+                    try {
+                        PaytrailPaymentContext context = paymentPaytrailService.buildPaytrailContext(message.getNamespace(), message.getMerchantId());
+                        PaytrailPaymentMitChargeSuccessResponse mitCharge = paymentPaytrailService.createMitCharge(context, payment.getPaymentId(), message);
+                        onlinePaymentService.setPaytrailTransactionId(payment.getPaymentId(), mitCharge.getTransactionId());
```
Comment (severikupari1) [inline paymentapi/src/main/java/fi/hel/verkkokauppa/payment/api/PaymentAdminController.java:121]: Should add signature validation

File: paymentapi/src/test/java/fi/hel/verkkokauppa/payment/api/PaymentAdminControllerTest.java [hunk 2]
```diff
@@ -377,6 +386,42 @@ public void whenGetPaymentMethodsThenReturnMoreThanZeroWithStatus200() {
...
+                    .cardLastFourDigits("0354")
+                    .orderItemId(UUIDGenerator.generateType4UUID().toString())
+                    .vatPercentage("0")
+                    .productName("productName")
+                    .productQuantity("1")
+                    .isSubscriptionRenewalOrder(true)
+                    .subscriptionId(UUIDGenerator.generateType4UUID().toString())
+                    .userId("user")
+                    .paymentGateway(fi.hel.verkkokauppa.common.constants.PaymentGatewayEnum.PAYTRAIL)
+                    .merchantId("0_YrloUBVL9Yu6dzAXua")
```
Comment (severikupari1) [inline paymentapi/src/test/java/fi/hel/verkkokauppa/payment/api/PaymentAdminControllerTest.java:415]: you could add https://github.com/City-of-Helsinki/verkkokauppa-core/blob/9efac21409a36a22e3edbf53f330ea7f035bc9f2/paymentapi/src/test/java/fi/hel/verkkokauppa/payment/testing/BaseFunctionalTest.java#L27 to fetch merchantid with namespace.

File: paymentapi/src/main/java/fi/hel/verkkokauppa/payment/service/OnlinePaymentService.java [hunk 3]
```diff
@@ -378,7 +384,7 @@ public PaymentCardInfoDto getPaymentCardInfo(String namespace, String orderId, S
         return paymentCardToken;
     }
 
-    public void triggerPaymentPaidEvent(Payment payment) {
+    public void triggerPaymentPaidEvent(Payment payment, PaymentCardInfoDto card) {
```
Comment (severikupari1) [inline paymentapi/src/main/java/fi/hel/verkkokauppa/payment/service/OnlinePaymentService.java:387]: Could you just make new function with new parameters and leave old as is?

File: paymentapi/src/main/java/fi/hel/verkkokauppa/payment/mapper/PaytrailCreatePaymentPayloadMapper.java [hunk 4]
```diff
@@ -11,10 +11,10 @@
  * > Map create payment payload to create mit charge payload and back
  */
 @Component
-public class PaytrailCreatePaymentPayloadMapper extends AbstractModelMapper<CreatePaymentPayload, CreateMitChargePayload> {
+public class PaytrailCreatePaymentPayloadMapper extends AbstractModelMapper<CreatePaymentPayload, PaytrailPaymentCreateMitChargeRequest.CreateMitChargePayload> {
```
Comment (severikupari1) [inline paymentapi/src/main/java/fi/hel/verkkokauppa/payment/mapper/PaytrailCreatePaymentPayloadMapper.java:14]: these can be changed back

File: paymentapi/src/main/java/fi/hel/verkkokauppa/payment/api/PaymentAdminController.java [hunk 5]
```diff
@@ -100,6 +102,45 @@ public ResponseEntity<List<Payment>> getPayments(@RequestParam(value = "orderId"
...
+    public ResponseEntity<Void> orderCreatedEventCallbackPaytrail(@RequestBody OrderMessage message) {
+        log.debug("orderCreatedEventCallback OrderMessage: {}", message);
+        try {
+            log.debug("payment-api received ORDER_CREATED event for orderId: " + message.getOrderId());
+
+            Payment existingPayment = onlinePaymentService.getPaymentForOrder(message.getOrderId());
+            if (existingPayment != null && PaymentStatus.PAID_ONLINE.equals(existingPayment.getStatus())) {
+                log.warn("paid payment exists, not creating new payment for orderId: " + message.getOrderId());
+            } else {
+                if (Boolean.TRUE.equals(message.getIsSubscriptionRenewalOrder()) && message.getCardToken() != null && message.getCardExpYear() != null && message.getCardExpMonth() != null && message.getCardLastFourDigits() != null) {
```
Comment (severikupari1) [inline paymentapi/src/main/java/fi/hel/verkkokauppa/payment/api/PaymentAdminController.java:115]: write some kind of helper function etc to check null values.

**Review by severikupari1 — APPROVED** [review]

**Pull request merged**